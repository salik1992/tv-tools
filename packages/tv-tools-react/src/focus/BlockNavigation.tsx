import type { PropsWithChildren } from 'react';
import { FocusContext, useFocusContainer } from '../focus';

function markAsProcessed() {
	return true;
}

/**
 * BlockNavigation is a component that prevents navigation to escape from its children.
 * Useful for Modal, VirtualKeyboard, etc.
 */
export const BlockNavigation = ({ children }: PropsWithChildren) => {
	const { focusContextValue, useOnLeft, useOnUp, useOnDown, useOnRight } =
		useFocusContainer();

	useOnLeft(markAsProcessed);
	useOnRight(markAsProcessed);
	useOnUp(markAsProcessed);
	useOnDown(markAsProcessed);

	return (
		<FocusContext.Provider value={focusContextValue}>
			{children}
		</FocusContext.Provider>
	);
};
