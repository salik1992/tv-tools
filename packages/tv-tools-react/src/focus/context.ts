import { createContext } from 'react';
import { noop } from '@salik1992/tv-tools/utils/noop';
import type { Focus } from './types';

export const FocusContext = createContext<Focus>({
	addChild: noop,
});
