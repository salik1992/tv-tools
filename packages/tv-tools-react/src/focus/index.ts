export { FocusContext } from './context';
export { FocusProvider } from './FocusProvider';
export { HorizontalFocus } from './HorizontalFocus';
export { Interactable } from './Interactable';
export { VerticalFocus } from './VerticalFocus';
export { FocusTable, FocusTableRenderComponents } from './FocusTable';
export { BlockNavigation } from './BlockNavigation';
export { useFocusContainer } from './useFocusContainer';
export { useFocusManager } from './useFocusManager';
export { useKeySequenceListener } from './useKeySequenceListener';
