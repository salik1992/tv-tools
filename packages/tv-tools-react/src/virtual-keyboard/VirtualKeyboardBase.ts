import { VirtualKeyboard } from '@salik1992/tv-tools/virtual-keyboard';

/**
 * VirtualKeyboardBase is extending the VirtualKeyboard class from tv-tools and
 * it overrides setInputValue to work with React where the value is edited on input
 * element and then the "input" event is fired.
 */
export class VirtualKeyboardBase extends VirtualKeyboard {
	protected override setInputValue(value: string) {
		if (this.input) {
			// React will override the value of the input element with its internal state
			// if we use just this.input.value = value;
			// https://stackoverflow.com/a/46012210
			const nativeInputSetter = Object.getOwnPropertyDescriptor(
				HTMLInputElement.prototype,
				'value',
			)?.set;
			if (nativeInputSetter) {
				nativeInputSetter.call(this.input, value);
			} else {
				this.input.value = value;
			}
		}
	}
}
