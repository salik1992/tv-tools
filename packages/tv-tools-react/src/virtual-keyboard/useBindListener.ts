import { useEffect } from 'react';
import type {
	VirtualKeyboard,
	VirtualKeyboardEvents,
} from '@salik1992/tv-tools/virtual-keyboard';

export function useBindListener<EventName extends keyof VirtualKeyboardEvents>(
	eventName: EventName,
	keyboard: VirtualKeyboard,
	callback:
		| (VirtualKeyboardEvents[typeof eventName] extends never
				? () => void
				: (payload: VirtualKeyboardEvents[typeof eventName]) => void)
		| undefined,
) {
	useEffect(() => {
		if (typeof callback === 'function') {
			keyboard.addEventListener(eventName, callback);
		}
		return () => {
			if (typeof callback === 'function') {
				keyboard.removeEventListener(eventName, callback);
			}
		};
	}, [keyboard, eventName, callback]);
}
