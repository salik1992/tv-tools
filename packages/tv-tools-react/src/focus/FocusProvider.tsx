import type { ComponentProps, FocusEvent, KeyboardEvent } from 'react';
import { useCallback, useMemo } from 'react';
import { isDirectional } from '@salik1992/tv-tools/control';
import { FocusManager } from '@salik1992/tv-tools/focus';
import { noop } from '@salik1992/tv-tools/utils/noop';
import { FocusContext } from './context';

/**
 * The main root component for the focus management.
 * It creates a div element that is used for listening to key events in the manager and
 * it also creates the focus manager instance for the application which is available through
 * the FocusContext or useFocusManager hook.
 * Use at the root of your application.
 * @prop alwaysPreventNavigationalEvents - Use this to prevent TV use their own navigation.
 * @prop ...props - Any other props that you want to pass to the wrapping div element.
 */
export const FocusProvider = ({
	children,
	alwaysPreventNavigationalEvents,
	onFocus: onOuterFocus,
	onKeyDownCapture: onOuterKeyDownCapture,
	onKeyDown: onOuterKeyDown,
	onKeyUpCapture: onOuterKeyUpCapture,
	onKeyUp: onOuterKeyUp,
	...props
}: ComponentProps<'div'> & {
	/** Use this to prevent TV use their own navigation. */
	alwaysPreventNavigationalEvents?: boolean;
}) => {
	const focusManager = useMemo(() => new FocusManager(), []);

	const onFocus = useCallback(
		(event: FocusEvent<HTMLDivElement>) => {
			focusManager.handleFocusEvent(event.nativeEvent);
			onOuterFocus?.(event);
		},
		[focusManager, onOuterFocus],
	);

	const keyDownCapture = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keydown', 'capture', event);
			onOuterKeyDownCapture?.(event);
		},
		[focusManager, onOuterKeyDownCapture],
	);

	const keyDownBubble = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keydown', 'bubble', event);
			if (
				alwaysPreventNavigationalEvents &&
				isDirectional(event) &&
				document.activeElement?.tagName !== 'INPUT'
			) {
				event.preventDefault();
				event.stopPropagation();
			} else {
				onOuterKeyDown?.(event);
			}
		},
		[focusManager, alwaysPreventNavigationalEvents, onOuterKeyDown],
	);

	const keyUpCapture = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keyup', 'capture', event);
			onOuterKeyUpCapture?.(event);
		},
		[focusManager, onOuterKeyUpCapture],
	);

	const keyUpBubble = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keyup', 'bubble', event);
			if (
				alwaysPreventNavigationalEvents &&
				isDirectional(event) &&
				document.activeElement?.tagName !== 'INPUT'
			) {
				event.preventDefault();
				event.stopPropagation();
			} else {
				onOuterKeyUp?.(event);
			}
		},
		[focusManager, alwaysPreventNavigationalEvents, onOuterKeyUp],
	);

	const focusContextValue = useMemo(
		() => ({
			focusManager,
			addChild: noop,
		}),
		[focusManager],
	);

	return (
		<div
			{...props}
			onFocus={onFocus}
			onKeyDownCapture={keyDownCapture}
			onKeyDown={keyDownBubble}
			onKeyUpCapture={keyUpCapture}
			onKeyUp={keyUpBubble}
		>
			<FocusContext.Provider value={focusContextValue}>
				{children}
			</FocusContext.Provider>
		</div>
	);
};
