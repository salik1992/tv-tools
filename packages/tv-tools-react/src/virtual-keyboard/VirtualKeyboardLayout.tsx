import {
	type CSSProperties,
	type ComponentType,
	type PropsWithChildren,
	type RefObject,
	useEffect,
	useMemo,
	useState,
} from 'react';
import type { TableFocusContainer } from '@salik1992/tv-tools/focus';
import { type VirtualKeyboardLayouts } from '@salik1992/tv-tools/virtual-keyboard';
import { VirtualKeyboardBase } from './VirtualKeyboardBase';
import { bindListener } from './bindListener';

export const VirtualKeyboardLayout = ({
	layouts,
	container,
	Keyboard,
	FocusTr,
	FocusTd,
	onAddChar,
	onRemoveChar,
	onDone,
	inputRef,
}: {
	layouts: VirtualKeyboardLayouts;
	container: TableFocusContainer;
	Keyboard: ComponentType<
		PropsWithChildren<{ className?: string; style?: CSSProperties }>
	>;
	FocusTr: ComponentType<PropsWithChildren>;
	FocusTd: ComponentType<
		PropsWithChildren<{
			id?: string;
			colSpan?: number;
			rowSpan?: number;
			className?: string;
		}>
	>;
	onAddChar?: (char: string) => void;
	onRemoveChar?: () => void;
	onDone?: () => void;
	inputRef?: RefObject<HTMLInputElement | null>;
}) => {
	const keyboard = useMemo(
		() => new VirtualKeyboardBase(layouts, container),
		[container, layouts],
	);
	const [renderData, setRenderData] = useState(keyboard.getRenderData());

	bindListener('renderData', keyboard, setRenderData);
	bindListener('addChar', keyboard, onAddChar);
	bindListener('removeChar', keyboard, onRemoveChar);
	bindListener('done', keyboard, onDone);

	useEffect(() => {
		keyboard.assignInput(inputRef?.current ?? undefined);
	}, [keyboard, inputRef?.current]);

	const className = useMemo(() => {
		let className = 'virtual-keyboard';
		className += ` layout-${renderData.layoutName}`;
		if (renderData.effect !== undefined) {
			className += ` effect-${renderData.effect}`;
		}
		return className;
	}, [renderData.effect, renderData.layoutName]);

	return (
		<Keyboard className={className}>
			{renderData.layout.map((row, rowIndex) => (
				<FocusTr key={rowIndex}>
					{row.map((key) => {
						const { key: keyId, ...props } = key;
						const id = `${container.id}-vk-${keyId}`;
						return (
							<FocusTd
								key={id}
								id={id}
								className={`vk-${keyId}`}
								{...props}
							>
								{key.label}
							</FocusTd>
						);
					})}
				</FocusTr>
			))}
		</Keyboard>
	);
};
