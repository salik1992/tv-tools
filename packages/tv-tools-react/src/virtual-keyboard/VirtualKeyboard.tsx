import { type ComponentType, Fragment, type RefObject } from 'react';
import type { TableFocusContainer } from '@salik1992/tv-tools/focus';
import type { VirtualKeyboardLayouts } from '@salik1992/tv-tools/virtual-keyboard';
import { FocusTable, type FocusTableRenderComponents } from '../focus';
import { BlockNavigation } from '../utils/BlockNavigation';
import { VirtualKeyboardLayout } from './VirtualKeyboardLayout';

export const VirtualKeyboard = ({
	layouts,
	blockNavigation = false,
	Keyboard,
	Row,
	Key,
	onAddChar,
	onRemoveChar,
	onDone,
	inputRef,
}: {
	layouts: VirtualKeyboardLayouts;
	Keyboard: ComponentType;
	Row: ComponentType;
	Key: ComponentType;
	blockNavigation?: boolean;
	onAddChar?: (char: string) => void;
	onRemoveChar?: () => void;
	onDone?: () => void;
	inputRef?: RefObject<HTMLInputElement | null>;
}) => {
	const Blocker = blockNavigation ? BlockNavigation : Fragment;
	return (
		<Blocker>
			<FocusTable
				TableComponent={Fragment}
				TrComponent={Row}
				TdComponent={Key}
			>
				{(
					{
						FocusTr,
						FocusTd,
					}: FocusTableRenderComponents<typeof Row, typeof Key>,
					container: TableFocusContainer,
				) => (
					<VirtualKeyboardLayout
						layouts={layouts}
						container={container}
						Keyboard={Keyboard}
						FocusTr={FocusTr}
						FocusTd={FocusTd}
						onAddChar={onAddChar}
						onRemoveChar={onRemoveChar}
						onDone={onDone}
						inputRef={inputRef}
					/>
				)}
			</FocusTable>
		</Blocker>
	);
};
