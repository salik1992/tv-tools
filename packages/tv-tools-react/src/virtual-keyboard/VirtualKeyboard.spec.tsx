import { PropsWithChildren } from 'react';
import { act, render } from '@testing-library/react';
import { ENTER } from '@salik1992/tv-tools/control';
import { backspace, done, layout } from '@salik1992/tv-tools/virtual-keyboard';
import { FocusRoot, Interactable } from '../focus';
import { VirtualKeyboard } from './VirtualKeyboard';

jest.useFakeTimers();

const LAYOUTS = {
	base: {
		initialKey: 'a',
		keys: [
			['a', 'b', 'c', 'd', 'e'],
			['f', 'g', 'h', 'i', 'j'],
			['k', 'l', 'm', 'n', 'o'],
			['p', 'q', 'r', 's', 't'],
			['u', 'v', 'w', 'x', 'y'],
			[
				'z',
				{ key: 'space', char: ' ' },
				{ key: 'dot', char: '.' },
				{ key: 'backspace', label: 'DEL', action: backspace },
				{ key: 'done', label: 'DONE', action: done },
			],
		],
	},
};

const Keyboard = ({ children }: PropsWithChildren) => (
	<div className="keyboard">{children}</div>
);

const Row = ({ children }: PropsWithChildren) => (
	<div className="row">{children}</div>
);

describe('VirtualKeyboard', () => {
	it('should render the keyboard', () => {
		const { container, unmount } = render(
			<VirtualKeyboard
				id="keyboard"
				layouts={LAYOUTS}
				Keyboard={Keyboard}
				Row={Row}
				Key={Interactable}
			/>,
		);
		expect(container).toMatchSnapshot();
		unmount();
	});

	it('should receive info from keyboard', () => {
		const onAddChar = jest.fn();
		const onRemoveChar = jest.fn();
		const onDone = jest.fn();

		const { container, unmount } = render(
			<VirtualKeyboard
				id="keyboard"
				layouts={LAYOUTS}
				Keyboard={Keyboard}
				Row={Row}
				Key={Interactable}
				onAddChar={onAddChar}
				onRemoveChar={onRemoveChar}
				onDone={onDone}
			/>,
			{ wrapper: FocusRoot },
		);

		act(() => {
			const jKey = container.querySelector('#keyboard-vk-j');
			jKey?.dispatchEvent(ENTER.toKeyboardEvent('keydown'));
			jest.runAllTimers();
		});
		expect(onAddChar).toHaveBeenCalledWith('j');

		act(() => {
			const backspaceKey = container.querySelector(
				'#keyboard-vk-backspace',
			);
			backspaceKey?.dispatchEvent(ENTER.toKeyboardEvent('keydown'));
			jest.runAllTimers();
		});
		expect(onRemoveChar).toHaveBeenCalled();

		act(() => {
			const doneKey = container.querySelector('#keyboard-vk-done');
			doneKey?.dispatchEvent(ENTER.toKeyboardEvent('keydown'));
			jest.runAllTimers();
		});
		expect(onDone).toHaveBeenCalled();

		unmount();
	});

	it('should render various layouts', () => {
		const { container, unmount } = render(
			<VirtualKeyboard
				id="keyboard"
				layouts={{
					abc: {
						initialKey: 'a',
						keys: [
							['a', 'b', 'c'],
							[
								{
									key: 'numbers',
									label: '123',
									action: layout('numbers'),
								},
							],
						],
					},
					numbers: {
						initialKey: '1',
						keys: [
							['1', '2', '3'],
							[
								{
									key: 'abc',
									label: 'ABC',
									action: layout('abc'),
								},
							],
						],
					},
				}}
				Keyboard={Keyboard}
				Row={Row}
				Key={Interactable}
			/>,
			{ wrapper: FocusRoot },
		);

		expect(container.innerHTML).toContain('123');
		expect(container.innerHTML).not.toContain('ABC');

		act(() => {
			const abcKey = container.querySelector('#keyboard-vk-numbers');
			abcKey?.dispatchEvent(ENTER.toKeyboardEvent('keydown'));
			jest.runAllTimers();
		});

		expect(container.innerHTML).not.toContain('123');
		expect(container.innerHTML).toContain('ABC');

		act(() => {
			const abcKey = container.querySelector('#keyboard-vk-abc');
			abcKey?.dispatchEvent(ENTER.toKeyboardEvent('keydown'));
			jest.runAllTimers();
		});

		expect(container.innerHTML).toContain('123');
		expect(container.innerHTML).not.toContain('ABC');

		unmount();
	});
});
