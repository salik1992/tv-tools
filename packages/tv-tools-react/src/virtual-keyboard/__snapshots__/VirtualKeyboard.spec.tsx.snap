// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualKeyboard should render the keyboard 1`] = `
<div>
  <div>
    <div
      class="keyboard"
    >
      <div
        class="row"
      >
        <div
          class="vk-a"
          id="keyboard-vk-a"
          label="a"
          tabindex="0"
        >
          a
        </div>
        <div
          class="vk-b"
          id="keyboard-vk-b"
          label="b"
          tabindex="0"
        >
          b
        </div>
        <div
          class="vk-c"
          id="keyboard-vk-c"
          label="c"
          tabindex="0"
        >
          c
        </div>
        <div
          class="vk-d"
          id="keyboard-vk-d"
          label="d"
          tabindex="0"
        >
          d
        </div>
        <div
          class="vk-e"
          id="keyboard-vk-e"
          label="e"
          tabindex="0"
        >
          e
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="vk-f"
          id="keyboard-vk-f"
          label="f"
          tabindex="0"
        >
          f
        </div>
        <div
          class="vk-g"
          id="keyboard-vk-g"
          label="g"
          tabindex="0"
        >
          g
        </div>
        <div
          class="vk-h"
          id="keyboard-vk-h"
          label="h"
          tabindex="0"
        >
          h
        </div>
        <div
          class="vk-i"
          id="keyboard-vk-i"
          label="i"
          tabindex="0"
        >
          i
        </div>
        <div
          class="vk-j"
          id="keyboard-vk-j"
          label="j"
          tabindex="0"
        >
          j
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="vk-k"
          id="keyboard-vk-k"
          label="k"
          tabindex="0"
        >
          k
        </div>
        <div
          class="vk-l"
          id="keyboard-vk-l"
          label="l"
          tabindex="0"
        >
          l
        </div>
        <div
          class="vk-m"
          id="keyboard-vk-m"
          label="m"
          tabindex="0"
        >
          m
        </div>
        <div
          class="vk-n"
          id="keyboard-vk-n"
          label="n"
          tabindex="0"
        >
          n
        </div>
        <div
          class="vk-o"
          id="keyboard-vk-o"
          label="o"
          tabindex="0"
        >
          o
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="vk-p"
          id="keyboard-vk-p"
          label="p"
          tabindex="0"
        >
          p
        </div>
        <div
          class="vk-q"
          id="keyboard-vk-q"
          label="q"
          tabindex="0"
        >
          q
        </div>
        <div
          class="vk-r"
          id="keyboard-vk-r"
          label="r"
          tabindex="0"
        >
          r
        </div>
        <div
          class="vk-s"
          id="keyboard-vk-s"
          label="s"
          tabindex="0"
        >
          s
        </div>
        <div
          class="vk-t"
          id="keyboard-vk-t"
          label="t"
          tabindex="0"
        >
          t
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="vk-u"
          id="keyboard-vk-u"
          label="u"
          tabindex="0"
        >
          u
        </div>
        <div
          class="vk-v"
          id="keyboard-vk-v"
          label="v"
          tabindex="0"
        >
          v
        </div>
        <div
          class="vk-w"
          id="keyboard-vk-w"
          label="w"
          tabindex="0"
        >
          w
        </div>
        <div
          class="vk-x"
          id="keyboard-vk-x"
          label="x"
          tabindex="0"
        >
          x
        </div>
        <div
          class="vk-y"
          id="keyboard-vk-y"
          label="y"
          tabindex="0"
        >
          y
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="vk-z"
          id="keyboard-vk-z"
          label="z"
          tabindex="0"
        >
          z
        </div>
        <div
          class="vk-space"
          id="keyboard-vk-space"
          label="space"
          tabindex="0"
        >
          space
        </div>
        <div
          class="vk-dot"
          id="keyboard-vk-dot"
          label="dot"
          tabindex="0"
        >
          dot
        </div>
        <div
          class="vk-backspace"
          id="keyboard-vk-backspace"
          label="DEL"
          tabindex="0"
        >
          DEL
        </div>
        <div
          class="vk-done"
          id="keyboard-vk-done"
          label="DONE"
          tabindex="0"
        >
          DONE
        </div>
      </div>
    </div>
  </div>
</div>
`;
